lib/Sema/CSGen.cpp

```diff
@@ -1480,6 +1526,47 @@ namespace {
     Type visitDynamicMemberRefExpr(DynamicMemberRefExpr *expr) {
       llvm_unreachable("Already typechecked");
     }
+
+    void setUnresolvedBaseType(UnresolvedMemberExpr *UME, Type ty) {
+      UnresolvedBaseTypes.insert({UME, ty});
+    }
+
+    Type addUnresolvedMemberChainConstraints(UnresolvedMemberExpr *base,
+                                             Expr *tail,
+                                        UnresolvedMemberChainResultExpr *result,
+                                             Type resultTy) {
+      assert(isa<UnresolvedMemberExpr>(tail) ||
+             isa<UnresolvedDotExpr>(tail) ||
+             isa<CallExpr>(tail) ||
+             isa<BindOptionalExpr>(tail) ||
+             isa<ForceValueExpr>(tail) ||
+             isa<SubscriptExpr>(tail) &&
+             "Unexpected expression at end of unresolved member chain");
+      // Copy any type variable options from the result of the tail member to
+      // the result of the entire chain.
+      unsigned additionalOptions = 0;
+      if (auto *tvt = resultTy->getAs<TypeVariableType>())
+        additionalOptions = tvt->getImpl().getRawOptions();
```

Comment:
Why do we need to do this?

Reply:
Some member types (namely, function calls) don't allow binding to lvalues. For single member chains, we currently pass in the options explicitly in `visitUnresolvedMemberExpr`, but that approach doesn't work if we are trying to separate the knowledge of when to insert unresolved member chain constraints from the `visit*` methods.

The other option would be to always pass `TVO_CanBindToLValue` to the result of a unresolved member chain, but I was worried that that would admit solutions that aren't actually valid (and that behavior is a larger departure from the current behavior for single-member chains). Open to other thoughts about how to handle this, though!

Reply:
I think it's okay to allow result type to be l-value, that's also how it's set up today, only base is not supposed to be lv-value (because it would be wrapped into a metatype).

Reply:
The current implementation actually allows lvalue for member accesses, but disallows lvalue for calls. I kept this logic in the interest of maintaining that behavior as closely as possible, but I could do a run of the tests where I just remove it and see if it causes any issues.

Reply:
Calls can't return l-value - result type variable doesn't allow that, only subscripts could and we should allow that too.